src/banjo_kazooie/asset.rs

```diff
@@ -521,11 +524,502 @@ pub struct LevelSetup{
+                        }
+
+                        todo!("Implement cmd {cmd}");
+                    }
+                }, 
+                _ => {
+                    // no-op
+                    todo!("Implement cmd {cmd}?");
+                },
+            }
+        }
+    }
+    
+    fn get_cubes_from_reader(reader: &mut LevelSetupReader) -> Vec<String> {
+        let mut out_cubes = vec![];
+
+        loop {
+            let cmd = reader.read_u8();
+            // println!("Cmd = {cmd}, offset = {offset}");
+
+            match cmd {
+                0 => {
+                    /*
+                    if (file_getNWords_ifExpected(file_ptr, 0, sp2C, 3)) {
+                        file_getNWords(file_ptr, sp2C, 3);
+                    */
+                    reader.read_n(6, |r| r.read_word());
+                },
+                1 => {
+                    return out_cubes;
+                },
+                2 => {
+                    /*
+                     !file_getNWords_ifExpected(file_ptr, 2, &sp2C, 3)
+                     */
+                    todo!("Cmd = 2");
+                    reader.read_n(3, |r| r.read_word());
+                },
+                3 => { // ->cube_fromFile
+                    let cube_type = reader.read_u8();
+                    let count: usize = reader.read_u8().into();
+
+                    let next_expected = match cube_type {
+                        0xA => 0xB,
+                        0x6 => 0x7,
+                        _ => panic!("Unsupported cude type ? cube_type = {cube_type}")
+                    };
+        
+                    println!("Reading {count} cubes (3)");
+                    
+                    let cube_byte_size = 20; // sizeof(NodeProp)
+                    let cubes = reader.read_if_expected(next_expected, |r| r.read_u8_n(count * cube_byte_size));
+
+                    if let Some(cubes) = cubes {
+                        cubes.chunks(cube_byte_size).for_each(|cube| {
+                            out_cubes.push(LevelSetupReader::u8s_to_string(cube));
+                        });
+                    }
+                },
+                8 => {
+                    // this "frees" the cube in the c code
+                    let count: usize = reader.read_u8().into();
+
+                    println!("Reading {count} cubes (8)");
+
+                    let cube_byte_size = 12; // sizeof(OtherNode)
+                    let cubes = reader.read_if_expected(9, |r| r.read_u8_n(count* cube_byte_size));
+
+                    if let Some(cubes) = cubes {
+                        cubes.chunks(cube_byte_size).for_each(|cube| {
+                            out_cubes.push(LevelSetupReader::u8s_to_string(cube));
+                        });
+                    }
+                },
+                _ => {
+                    todo!("Unknown cmd {cmd}");
+                },
+            }
+        }
+    }
+
+
+    fn build_map_hash_set() -> HashMap<usize, String> {
```

Comment:
note: This will not be part of the final PR. I just added this here for easier debugging when reading out level data